File: cpp/023MergeKSortedLists/023MergeKSortedLists.cpp
#include <sstream>
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

/*
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
*/


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        int K = lists.size();       
        if(K==0) return NULL;

        for(int step = 1; (1<<(step-1))<K; ++step){
            for(int k=0; k<K; k+=(1<<step)){
                if(k+(1<<(step-1)) < K){
                    lists[k] = mergeTwoLists(lists[k], lists[k+(1<<(step-1))]);
                }
            }
        }
       

        return lists[0];
    }

    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        if(l1==NULL) return l2;
        if(l2==NULL) return l1;

        ListNode *head, *cur;
        if(l1->val<l2->val){
            cur = head = l1;
            l1=l1->next;
        }else{
            cur = head = l2;
            l2=l2->next;
        }

        while(l1!=NULL&&l2!=NULL){
            if(l1->val<l2->val){
                cur->next = l1; 
                l1=l1->next;
            }else{
                cur->next = l2;
                l2=l2->next;
            }
            cur = cur->next;
        }

        while(l1!=NULL){
            cur->next = l1; 
            l1=l1->next;
            cur = cur->next;
        }
        while(l2!=NULL){
            cur->next = l2;
            l2=l2->next;
            cur = cur->next;
        }
        cur->next=NULL;

        return head;
    }
};



int main(){
	Solution s;

	return 0;
} 

File: src/023MergeKSortedLists/023MergeKSortedLists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

use std::collections::BinaryHeap;
use std::cmp::Reverse;
use std::cmp::Ordering;

impl Ord for ListNode {
    fn cmp(&self, other: &ListNode) -> Ordering {
        self.val.cmp(&other.val)
    }
}

impl PartialOrd for ListNode {
    fn partial_cmp(&self, other: &ListNode) -> Option<Ordering> {
        Some(self.val.cmp(&other.val))
    }
}

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut min_heap = BinaryHeap::new();
        for list in lists{
            if let Some(ln) = list{
                min_heap.push(Reverse(ln));
            }
        }
        
        let mut dummy = ListNode::new(0);
        let mut p = &mut dummy;
        while !min_heap.is_empty() {
            //https://doc.rust-lang.org/src/core/cmp.rs.html#423
            //pub struct Reverse<T>(pub T);
            //Reverse is tuple-style struct, therefore, 
            //use .0 for access the internal data
            let mut list = min_heap.pop().unwrap().0;
            if let Some(ln) = list.next.take(){
                min_heap.push(Reverse(ln));
            } 
            p.next = Some(list);
            p = p.next.as_mut().unwrap();
        }
        
        dummy.next
    }
}
